[MMB-219] Update docs (#113)
ably-labs/spaces: docs/class-definitions.md

})
```

# Members

Handles members within a space.

## Methods

### subscribe

Listen to member events for the space. See [EventEmitter](/docs/usage.md#event-emitters) for overloading usage.

Available events:

- #### **update**

  Listen to all updates for members in a space.

  ```ts
  space.members.subscribe('update', (member: SpaceMember) => {});
  ```

  This is the same as not specifying the event:

  ```ts
  space.members.subscribe((member: SpaceMember) => {})

  ```

  Triggers on:
  - presence updates ([`enter`, `leave`, `update` events](https://ably.com/docs/presence-occupancy/presence?lang=javascript))
  - [location updates](#locationupdate)

  The argument supplied to the callback is the [SpaceMember](#spacemember) object representing the member that triggered the event.

- #### **enter**
  Listen to enter events of members.

  ```ts
  space.members.subscribe('enter', (member: SpaceMember) => {})
  ```
  The argument supplied to the callback is a [SpaceMember](#spacemember) object representing the member entering the space.

- #### **leave**

>   Listen to leave events of members. Note that the leave event will only fire once the [offlineTimeout](#spaceoptions) has passed.

Reviewer: Yeah, that's my understanding too. Can you please check this in the implementation @surminus?

>   Listen to leave events of members. Note that the leave event will only fire once the [offlineTimeout](#spaceoptions) has passed.

Author: > Can you please check this in the implementation

Seems to be but honestly not entirely confident that I understand the code

  ```ts
  space.members.subscribe('leave', (member: SpaceMember) => {})
  ```

  The argument supplied to the callback is a [SpaceMember](#spacemember) object representing the member leaving the space.

### unsubscribe

Remove all event listeners, all event listeners for an event, or specific listeners. See [EventEmitter](/docs/usage.md#event-emitters) for detailed usage.

```ts
// Unsubscribe from all events
space.members.unsubscribe();

// Unsubscribe from enter events
space.members.unsubscribe('enter');

// Unsubscribe from leave events
space.members.unsubscribe('leave');
```

### getSelf

Gets the [SpaceMember](#spacemember) object which relates to the local connection. Will return `undefined` if the client hasn't entered the space yet.

```ts
type getSelf = () => <SpaceMember | undefined>;
```

> Example:

Reviewer: Nice one with the examples!

})
```

# Members

Handles members within a space.

## Methods

### subscribe

Listen to member events for the space. See [EventEmitter](/docs/usage.md#event-emitters) for overloading usage.

Available events:

- #### **update**

  Listen to all updates for members in a space.

  ```ts
  space.members.subscribe('update', (member: SpaceMember) => {});
  ```

  This is the same as not specifying the event:

  ```ts
  space.members.subscribe((member: SpaceMember) => {})

  ```

  Triggers on:

>   - presence updates ([`enter`, `leave`, `remove`, `update` events](https://ably.com/docs/presence-occupancy/presence?lang=javascript))

Reviewer 2: There is no presence `remove` event, the `remove` event is a Spaces SDK specific event emitted 120s after a `leave` event without a subsequent `enter` or `update` for the same member, so I think it's worth its own bullet point.

Reviewer 2: Saying that, I see this is in the section describing the behaviour of the members `update` event, which is emitted [when a member's profile data changes](https://github.com/ably-labs/spaces/blob/main/src/Members.ts#L55-L60),  so the description I give above belongs in the section describing the `remove` event, not the `update` event, and this section should say the `update` event is triggered on profile data changes.

Reviewer 2: > I understood it to be subscribing to all updates

So `space.subscribe('update', ...)` subscribes to all updates within a space (i.e. whether its a profile data change, a location change, or a lock change, basically any `enter`, `update`, or `leave` presence event).

`space.members.subscribe('update', ...)` is however _just_ for profile data changes, with location changes emitted by `space.locations.subscribe` and lock changes by `space.locks.subscribe`.

Given your confusion though, perhaps this is not a particularly intuitive way to divide the events across the different classes (although I think something to discuss separately, let's just focus on describing the current behaviour here).

>   - [location updates](#locationupdate)

Reviewer 2: As per [this code comment](https://github.com/ably-labs/spaces/blob/main/src/Members.ts#L55-L56), the member `update` event does _not_ fire for location updates, only profile data updates.

- #### **leave**

  Listen to leave events of members. The leave event will be issued when a member calls `space.leave()` or is disconnected.

  ```ts
  space.members.subscribe('leave', (member: SpaceMember) => {})
  ```

  The argument supplied to the callback is a [SpaceMember](#spacemember) object representing the member leaving the space.

- #### **remove**

  Listen to remove events of members. The remove event will be issued when the [offlineTimeout](#spaceoptions) has passed.

  ```ts
  space.members.subscribe('remove', (member: SpaceMember) => {})
  ```

  The argument supplied to the callback is a [SpaceMember](#spacemember) object representing the member removed from the space.

### unsubscribe

Remove all event listeners, all event listeners for an event, or specific listeners. See [EventEmitter](/docs/usage.md#event-emitters) for detailed usage.

```ts
// Unsubscribe from all events
space.members.unsubscribe();

// Unsubscribe from enter events
space.members.unsubscribe('enter');

// Unsubscribe from leave events
space.members.unsubscribe('leave');
```

### getSelf

> Gets the [SpaceMember](#spacemember) object which relates to the local connection. Will return `undefined` if the client hasn't entered the space yet.

Reviewer 2: ```suggestion
Returns a Promise which resolves to the [SpaceMember](#spacemember) object relating to the local connection. Will resolve to `undefined` if the client hasn't entered the space yet.
```

```ts

> type getSelf = () => <SpaceMember | undefined>;

Reviewer 2: ```suggestion
type getSelf = () => Promise<SpaceMember | undefined>;
```

```

Example:

```ts

> const myMember = space.members.getSelf();

Reviewer 2: ```suggestion
const myMember = await space.members.getSelf();
```

```

### getAll

> Returns an array of all [SpaceMember](#spacemember) objects (members) currently in the space, including any who have left and not yet timed out. (_see: [offlineTimeout](#spaceoptions)_)

Reviewer 2: ```suggestion
Returns a Promise which resolves to an array of all [SpaceMember](#spacemember) objects (members) currently in the space, including any who have left and not yet timed out. (_see: [offlineTimeout](#spaceoptions)_)
```

```ts

> type getAll = () => SpaceMember[];

Reviewer 2: ```suggestion
type getAll = () => Promise<SpaceMember[]>;
```

```

Example:

```ts

> const allMembers = space.members.getAll();

Reviewer 2: ```suggestion
const allMembers = await space.members.getAll();
```

```

### getOthers

> Returns an array of all [SpaceMember](#spacemember) objects (members) currently in the space, excluding your own member object.

Reviewer 2: ```suggestion
Returns a Promise which resolves to an array of all [SpaceMember](#spacemember) objects (members) currently in the space, excluding your own member object.
```

```ts
type set = (update: Location) => void;
```

> ### subscribe

Reviewer: Can we move subscribe before set please in terms of the order of the methods? Thanks

## Methods

### subscribe

Listen to events for locations. See [EventEmitter](/docs/usage.md#event-emitters) for overloading usage.

Available events:

> - #### **locationUpdate**

Reviewer 2: ```suggestion
- #### **update**
```

(see https://github.com/ably-labs/spaces/blob/main/src/Locations.ts#L15).

>   Fires when a member updates their location. The argument supplied to the event listener is an [LocationUpdate](#locationupdate-1).

Reviewer 2: ```suggestion
  Fires when a member updates their location. The argument supplied to the event listener is a [LocationUpdate](#locationupdate-1).
```

  ```ts

>   space.locations.subscribe('locationUpdate', (locationUpdate: LocationUpdate) => {});

Reviewer 2: ```suggestion
  space.locations.subscribe('update', (locationUpdate: LocationUpdate) => {});
```

  ```

### set

Set your current location. [Location](#location-1) can be any JSON-serializable object. Emits a [locationUpdate](#locationupdate) event to all connected clients in this space.

```ts

> type set = (update: Location) => void;

Reviewer 2: ```suggestion
type set = (update: Location) => Promise<void>;
```

```

### unsubscribe

Remove all event listeners, all event listeners for an event, or specific listeners. See [EventEmitter](/docs/usage.md#event-emitters) for detailed usage.

```ts

> space.locations.unsubscribe('locationUpdate');

Reviewer 2: ```suggestion
space.locations.unsubscribe('update');
```

```

### getSelf

Get location for self.

```ts

> type getSelf = () => <Location | undefined>;

Reviewer 2: ```suggestion
type getSelf = () => Promise<Location | undefined>;
```

```

Example:

```ts
const myLocation = await space.locations.getSelf();
```

### getAll

Get location for all members.

```ts

> type getAll = () => Record<ConnectionId, Location>;

Reviewer 2: ```suggestion
type getAll = () => Promise<Record<ConnectionId, Location>>;
```

```

Example:

```ts
const allLocations = await space.locations.getAll();
```

### getOthers

Get location for other members

```ts

> type getOthers = () => Record<ConnectionId, Location>;

Reviewer 2: ```suggestion
type getOthers = () => Promise<Record<ConnectionId, Location>>;
```

>   space.cursors.subscribe('cursorsUpdate', (cursorUpdate: CursorUpdate) => {});

Reviewer: Can we please move subscribe to the top of the list after `set` but before the `get..` methods?

});
```

> ### subscribe

Reviewer: Similar to locations, can we please move subscribe before set wrt order?

> Listen to `CursorUpdate` events. See [EventEmitter](/docs/usage.md#event-emitters) for overloading usage.

Reviewer: ```suggestion
Listen to `CursorUpdate` events. See [EventEmitter](/docs/usage.md#event-emitters) for overloaded usage.
```

Available events:

> - #### **cursorsUpdate**

Reviewer: I don't think this is the event name anymore. Isn't it just `update` now?

const selfPosition = space.cursors.getSelf();
```

### getAll

Get the last CursorUpdate for each connection.

```ts
type getAll = () => Record<ConnectionId, CursorUpdate>;
```

Example:

```ts

> const lastPositions = space.cursors.getAll();

Reviewer: ```suggestion
const allLatestPositions = space.cursors.getAll();
```
